Devices: Desktop: dumb component

cjb opened the pull request:
@keybase/react-hackers 

![deviceslist](https://cloud.githubusercontent.com/assets/21217/14859200/1a349304-0c71-11e6-897b-0e64e0be0eba.gif)

Lots of questions about this PR!  Like when to use ES6 classes (currently, any time I have React state) and how to do `this.highlightOn = this.highlightOn.bind(this)` more cleanly, and whether any of this should be pulled up a level.  I'll probably sit down with Nojima and clean this up.

Please take a look, thanks!

chrisnojima commented on shared/devices/dumb.desktop.js:
```diff
@@ -0,0 +1,72 @@
+import Devices from './index.render.desktop'
+import type {Device} from '../constants/types/flow-types'
+
+const dev1: Device = {
```
the props we send to our components should be our idealized props and not the output of an rpc. the action/reducer should transform it from this into the clean props

chrisnojima commented on shared/devices/index.render.desktop.js:
```diff
@@ -1,160 +1,193 @@
+    this.state = {
+      expanded: false
+    }
+    this.toggleHeader = this.toggleHeader.bind(this)
```
we haven't been doing this up until now. there's some performance pluses but we haven't been caching these calls. intentional? also you can just use an arrow instead of bind

chrisnojima commented on shared/devices/index.render.desktop.js:
```diff
@@ -1,160 +1,193 @@
-        </div>
-      </div>
-    )
+  toggleHeader () {
```
we should start marking private functions with a _ prefix

chrisnojima commented on shared/devices/index.render.desktop.js:
```diff
@@ -1,160 +1,193 @@
-          {this.props.devices && this.props.devices.map(device => this.renderDevice(device))}
-        </div>
-      </div>
+      <Box key={this.props.device.name} onMouseOver={this.highlightOn} onMouseOut={this.highlightOff} style={{...stylesCommonRow, backgroundColor: this.props.revoked ? globalColors.lightGrey : globalColors.white}}>
```
previously we've been using css pseudo selectors to do this kind of thing. this approach is fine, just pointing it out

chrisnojima commented on shared/devices/index.render.desktop.js:
```diff
@@ -1,160 +1,193 @@
+    this.highlightOff = this.highlightOff.bind(this)
+  }
+
+  highlightOn (e) {
```
why is this 2 functions and not just one that takes in a highlighted param?

chrisnojima commented on shared/devices/index.render.desktop.js:
```diff
@@ -1,160 +1,193 @@
+  return (
+    <RevokedHeader>
+      <RevokedDescription/>
+      {revokedDevices.map(device => <DeviceRow device={device} revoked/>)}
```
you'll likely need a key here

chrisnojima commented on shared/devices/index.render.desktop.js:
```diff
@@ -1,160 +1,193 @@
+const Render = ({devices, revokedDevices, waitingForServer, addNewDevice, showRemoveDevicePage, showExistingDevicePage}: Props) => {
+  return (
+    <Box style={stylesContainer}>
+      {DeviceHeader(addNewDevice)}
```
you can use the stateless component form here
```
<DeviceHeader addNewDevice={addNewDevice}/>
```

chrisnojima replied:
and change the function to be const DeviceHeader = ({addNewDevice}) => { since it converts the props to an object and calls it

chrisnojima commented on shared/devices/index.render.desktop.js:
```diff
@@ -1,160 +1,193 @@
+    <Box style={stylesContainer}>
+      {DeviceHeader(addNewDevice)}
+      {devices && devices.map(device => <DeviceRow device={device} showRemoveDevicePage={showRemoveDevicePage} showExistingDevicePage={showExistingDevicePage}/>)}
+      {revokedDevices && RevokedDevices(revokedDevices)}
```
same here

cjb commented:
Thanks!  Fixed: fat arrow handlers, internal functions, added `key`s, used stateless component invocation form.

Not fixed:
* I'd love to use CSS pseudo-elements instead but can't figure out how you'd do it with inline styles without adding something like Radium.
* Device payload is still the verbatim RPC payload, not yet sure which fields we need (perhaps all of them?), can come back to it after the integration's done.

chrisnojima commented:
you can inject the styles into the components directly, see https://github.com/keybase/client/blob/0a5beec86ad25cea74d2e0df346dbba092196998/shared/devices/existing-device/index.render.desktop.js#L18
the idea with the props is its driven by the component so its more that. what do you need to draw the main window and what should it be called. let's discuss 2moro

MarcoPolo commented on shared/devices/index.render.desktop.js:
```diff
@@ -1,160 +1,186 @@
 
-import moment from 'moment'
-import commonStyles from '../styles/common'
+class RevokedHeader extends Component {
```
prefer this:

```js
type State = {expanded: boolean}

class RevokedHeader extends Component<void, Props, State> {
  state: State;
  ...
}
```

because if you use the Component with generics like `Component<DefaultProps, Props, State>` then you can do really cool stuff like: [PropsOf](https://github.com/keybase/client/blob/master/shared/constants/types/more.js#L22) and have better types!

MarcoPolo commented on shared/devices/index.render.desktop.js:
```diff
@@ -1,160 +1,186 @@
-        </div>
-      </div>
-    )
+  constructor (props) {
```
Flow doesn't complain about needing to type props?

MarcoPolo commented on shared/devices/index.render.desktop.js:
```diff
@@ -1,160 +1,186 @@
-      return this.renderPaperKey(device)
-    } else {
-      console.warn('Unknown device type: ' + device.type)
+class DeviceRow extends Component {
```
Same prefer suggestion as above

MarcoPolo commented on shared/devices/index.render.desktop.js:
```diff
@@ -1,160 +1,186 @@
-  showRemoveDevicePage: React.PropTypes.func.isRequired,
-  showExistingDevicePage: React.PropTypes.func.isRequired,
-  showGenPaperKeyPage: React.PropTypes.func.isRequired
+const RevokedDescription = () => {
```
🐑 🐑  replace the `{ return (...) }` with `( ... )`

MarcoPolo commented:
Looks good! I'd suggest spending a little time trying to do this in CSS like Nojima's example and sprucing up the types.

otherwise 👍

cjb commented:
Added flow types and used pseudo-CSS selectors for hover, turning DeviceRow into another stateless component!  Thanks both.  I'll merge in now.  (Still going to work on idealized dumb prop reducer.)
